server/data/data-library/src/main/java/org/apache/james/user/lib/model/Algorithm.java

```diff
@@ -27,63 +27,62 @@
 public class Algorithm {
 
     public enum HashingMode {
-        LEGACY(LEGACY_FACTORY),
-        DEFAULT(DEFAULT_FACTORY);
+        PLAIN,
+        SALTED,
+        LEGACY,
+        LEGACY_SALTED;
 
         public static HashingMode parse(String value) {
             return Arrays.stream(values())
                 .filter(aValue -> value.equalsIgnoreCase(aValue.toString()))
                 .findFirst()
-                .orElseThrow(() -> new IllegalArgumentException("Unsuported value for HashingMode: " + value + ". Should be one of " + ImmutableList.copyOf(values())));
-        }
-
-        private final Factory factory;
-
-        HashingMode(Factory factory) {
-            this.factory = factory;
-        }
-
-        public Factory getFactory() {
-            return factory;
+                .orElseThrow(() -> new IllegalArgumentException("Unsupported value for HashingMode: " + value + ". Should be one of " + ImmutableList.copyOf(values())));
         }
     }
 
-    public interface Factory {
-        Algorithm of(String rawValue);
+    public static Algorithm of(String algorithmName) {
+        return Algorithm.of(algorithmName, HashingMode.PLAIN);
     }
 
-    public static class LegacyFactory implements Factory {
-        @Override
-        public Algorithm of(String rawValue) {
-            return new Algorithm(rawValue, LEGACY);
-        }
+    public static Algorithm of(String algorithmName, String fallbackHashingMode) {
+        return Algorithm.of(algorithmName, HashingMode.parse(fallbackHashingMode));
     }
 
-    public static class DefaultFactory implements Factory {
-        @Override
-        public Algorithm of(String rawValue) {
-            return new Algorithm(rawValue, !LEGACY);
+    public static Algorithm of(String algorithmName, HashingMode fallbackHashingMode) {
+        String[] spec = algorithmName.split("/");
+        if (spec.length == 1) {
+            return new Algorithm(algorithmName, fallbackHashingMode);
+        } else {
+            return new Algorithm(spec[0], HashingMode.parse(spec[1]));
         }
```

Comment:
String.split uses internally REGEXs.

I would prefer using google Splitter here:

```
        List<String> spec = Splitter.on('/').splitToList(algorithmName);
        if (spec.size() == 1) {
            return new Algorithm(algorithmName, fallbackHashingMode);
        } else {
            return new Algorithm(spec.get(0), HashingMode.parse(spec.get(1)));
        }
```

Reply:
Sure, I wasn't aware that we have this dependecy available here.